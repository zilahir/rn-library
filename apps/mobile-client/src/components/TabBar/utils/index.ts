export { default as icons } from "./icons";

export { onPress } from "./navigations";
