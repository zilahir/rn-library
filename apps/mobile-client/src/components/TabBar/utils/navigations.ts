import { BottomTabBarProps } from "@react-navigation/bottom-tabs";

interface INavigationProps {
  navigation: BottomTabBarProps["navigation"];
  targetKey: string;
}

/**
 *
 * @param root0
 * @param root0.navigation
 * @param root0.target
 * @param root0.isFocused
 * @param root0.targetKey
 * @param root0.targetName
 */
export function onPress({
  navigation,
  targetKey,
  isFocused,
  targetName,
}: INavigationProps & { isFocused: boolean; targetName: string }): void {
  const event = navigation.emit({
    type: "tabPress",
    target: targetKey,
    canPreventDefault: true,
  });

  if (!isFocused && !event.defaultPrevented) {
    navigation.navigate(targetName);
  }
}

/*
export function onLongPress({ navigation, targetKey }: INavigationProps): void {
  navigation.emit({
    type: "tabLongPress",
    target: targetKey,
  });
}
*/
