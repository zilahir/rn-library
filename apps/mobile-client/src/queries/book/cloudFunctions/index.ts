export { default as fetchAllBooks } from "./fetchAllBooks";
