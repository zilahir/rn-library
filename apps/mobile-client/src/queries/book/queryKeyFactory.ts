const bookQueryKeyFactory = {
  all: ["book"] as const,
  getBookById: (id: string) => [...bookQueryKeyFactory.all, id] as const,
  getBookByIsbn: (isbn: string) => [...bookQueryKeyFactory.all, isbn] as const,
};

export default bookQueryKeyFactory;
