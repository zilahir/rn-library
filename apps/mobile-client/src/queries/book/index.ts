import { useQuery, UseQueryResult } from "@tanstack/react-query";
import { Book } from "pikkukirjasto-types/types/book";

import bookQueryKeyFactory from "./queryKeyFactory"; // TODO: change this import
import { fetchAllBooks } from "./cloudFunctions"; // TODO: change this import

/**
 * @description React hook for fetching all books
 * @returns {UseQueryResult<Book[]>} React Query result object
 */
function useBooks(): UseQueryResult<Book[]> {
  const books = useQuery({
    queryKey: bookQueryKeyFactory.all,
    queryFn: fetchAllBooks,
    refetchOnWindowFocus: false,
    retry: false,
    initialData: [],
    keepPreviousData: true,
  });

  return books;
}

export default useBooks;
