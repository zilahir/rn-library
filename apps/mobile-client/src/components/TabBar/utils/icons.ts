import { Feather } from "@expo/vector-icons";
import { ComponentProps } from "react";
import { BottomTabParamList } from "@app/navigation/types";

const icons: Record<
  keyof BottomTabParamList,
  ComponentProps<typeof Feather>["name"]
> = {
  Home: "home",
  Profile: "user",
  Trending: "trending-up",
  Library: "book-open",
  Borrow: "book",
};

export default icons;
