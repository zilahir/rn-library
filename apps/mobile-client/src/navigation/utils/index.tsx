import { ComponentProps } from "react";
import { BottomTabParamList } from "@app/navigation/types";
import HomeScreen from "@app/screens/HomeScreen";
import ProfileScreen from "@app/screens/ProfileScreen";
import TrendingScreen from "@app/screens/TrendingScreen";
import LibraryScreen from "@app/screens/LibraryScreen";

type ScreenName = keyof BottomTabParamList;

interface BottomTabScreen {
  label?: ScreenName;
  name: ScreenName;
  tabComponent: ComponentProps<any>;
  hideLabel?: boolean;
  isFloating?: boolean;
}

interface IBottomTabScreens {
  screens: BottomTabScreen[];
  getBottomTabBarTabs: () => BottomTabScreen[];
  getNumberOfTabBarTabs: () => number;
}

export const bottomTabScreens: IBottomTabScreens = {
  screens: [
    {
      label: "Home",
      name: "Home",
      tabComponent: HomeScreen,
    },

    {
      label: "Profile",
      name: "Profile",
      tabComponent: ProfileScreen,
    },
    {
      label: "Trending",
      name: "Trending",
      tabComponent: TrendingScreen,
    },
    {
      label: "Library",
      name: "Library",
      tabComponent: LibraryScreen,
    },
  ],
  getBottomTabBarTabs: (): BottomTabScreen[] =>
    bottomTabScreens.screens.map((screen) => ({
      ...screen,
      hideLabel: !!screen.hideLabel,
      isFloating: !!screen.isFloating,
    })),

  getNumberOfTabBarTabs: (): number => bottomTabScreens.screens.length,
};
