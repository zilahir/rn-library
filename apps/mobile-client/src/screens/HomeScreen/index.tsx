import { ReactElement, useMemo } from "react";
import { View, StyleSheet, ScrollView } from "react-native";
import { random, sampleSize } from "lodash";
import { useTheme } from "@shopify/restyle";
import { useTranslation } from "react-i18next";
import { TouchableOpacity } from "react-native-gesture-handler";
import useBooks from "@app/queries/book";
import withSafeAreaView from "@app/components/WithSafeAreaView";
import MainLayout from "@app/layouts/MainLayout";
import Featured from "@app/components/Featured";
import { Theme } from "@app/theme";
import { RootTabScreenProps } from "@app/navigation/types";
import SectionTitle from "@app/components/common/SectionTitle";
import Badge from "@app/components/common/Badge";
import Card from "@app/components/common/Card";
import SearchBar from "@app/components/SearchBar";
import ScreenTitle from "@app/components/common/ScreenTitle";

import BookList from "./components/BookList";

const styles = ({
  marginVertical,
}: {
  marginVertical: number;
}): ReturnType<typeof StyleSheet.create> =>
  StyleSheet.create({
    bookListContainer: {
      marginVertical: marginVertical,
      flex: 1,
      justifyContent: "center",
    },

    allBookContainer: {
      justifyContent: "center",
      flex: 1,
      flexDirection: "column",
    },
  });

/**
 *
 * @param {object} root0 - props
 * @param {object} root0.navigation - navigation object
 * @param {object} root0.route - route object
 * @returns {ReactElement} - react component
 * @description - renders the home screen
 */
function HomeScreen({
  navigation,
  route,
}: RootTabScreenProps<"Library">): ReactElement {
  const { data: books = [] } = useBooks();
  const { spacing } = useTheme<Theme>();

  const style = styles({ marginVertical: spacing.s });

  const featuredBooks = useMemo(
    () => sampleSize(books, random(5, 11)),
    [books]
  );

  const { t } = useTranslation();

  return (
    <MainLayout>
      <ScreenTitle title={t("homeScreen.title")} />
      <ScrollView showsVerticalScrollIndicator={false} horizontal={false}>
        <View testID={books?.length.toString()}>
          <SearchBar route={route} navigation={navigation} />
          {books && books.length > 0 && (
            <>
              <View style={[style.bookListContainer]}>
                <SectionTitle title={t("libraryScreen.featured")} />
                <Featured navigation={navigation} books={featuredBooks} />
              </View>
              <View style={[style.bookListContainer]}>
                <SectionTitle title={t("libraryScreen.popular")} />
                <Card variant="featured">
                  <BookList
                    books={books.slice(0, 5)}
                    variant="horizontal"
                    navigation={navigation}
                    hasPagination={false}
                  />
                </Card>
              </View>
              <View style={[style.allBookContainer]}>
                <SectionTitle
                  action={
                    <TouchableOpacity
                      testID="library"
                      onPress={(): void => navigation.navigate("Library")}
                    >
                      <Badge
                        type="outlinedBadge"
                        text={t("libraryScreen.all")}
                      />
                    </TouchableOpacity>
                  }
                  title={t("libraryScreen.all")}
                />

                <BookList
                  books={books.slice(5, 10)}
                  variant="vertical"
                  navigation={navigation}
                />
              </View>
            </>
          )}
        </View>
      </ScrollView>
    </MainLayout>
  );
}

export default withSafeAreaView(HomeScreen);
