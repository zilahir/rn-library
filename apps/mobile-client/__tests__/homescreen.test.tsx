/* eslint-disable react/jsx-props-no-spreading */
// @ts-expect-error this is a mock
import mockRNDeviceInfo from "react-native-device-info/jest/react-native-device-info-mock";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { fireEvent } from "@testing-library/react-native";
import HomeScreen from "@app/screens/HomeScreen";
import * as hooks from "@app/queries/book";

import { render } from "./utils/wrapper";

const demoBooks = [
  {
    isbn: "demo isbn",
    author: "Terrell Stroman",
    title: "dedico adhuc tripudio",
    cover: "https://picsum.photos/seed/9YjBfGro/640/480",
  },
];

const props = {
  navigation: {
    navigate: jest.fn(),
  },
  route: {
    params: {},
  },
} as any;

jest.mock("react-native/Libraries/EventEmitter/NativeEventEmitter");
jest.mock("react-native-device-info", () => mockRNDeviceInfo);

describe("HomeScreen", () => {
  it("renders correctly", () => {
    jest.spyOn(hooks, "default").mockImplementation((): any => ({
      data: [...demoBooks],
      isLoading: false,
      initialData: [...demoBooks],
    }));
    const { toJSON, getByTestId } = render(
      <QueryClientProvider client={new QueryClient()}>
        <HomeScreen {...props} />
      </QueryClientProvider>
    );
    expect(toJSON()).toMatchSnapshot();

    const toLibraryButton = getByTestId("library");

    expect(toLibraryButton).toBeDefined();

    fireEvent.press(toLibraryButton);
    expect(props.navigation.navigate).toHaveBeenCalled();
  });
});
