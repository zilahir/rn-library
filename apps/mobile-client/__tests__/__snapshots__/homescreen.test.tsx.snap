// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders correctly 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "backgroundColor": "#fff",
        "borderBottomLeftRadius": 25,
        "borderBottomRightRadius": 25,
        "flex": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "transparent",
          "flex": 1,
          "marginHorizontal": 24,
          "marginVertical": 16,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "marginBottom": 32,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "rgba(51, 51, 51, 1)",
              "fontSize": 26,
              "fontWeight": "bold",
            },
          ]
        }
      >
        homeScreen.title
      </Text>
    </View>
    <RCTScrollView
      horizontal={false}
      showsVerticalScrollIndicator={false}
    >
      <View>
        <View
          testID="1"
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "marginBottom": 16,
                "opacity": 1,
              }
            }
            testID="search-bar"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "rgba(247, 247, 247, 1)",
                    "borderRadius": 10,
                    "display": "flex",
                    "flexDirection": "row",
                    "height": 40,
                    "paddingHorizontal": 8,
                  },
                ]
              }
            >
              <Text />
              <TextInput
                editable={false}
                placeholder="Search"
                placeholderTextColor="rgba(247, 247, 247, 1)"
                style={
                  [
                    {
                      "color": "#828282",
                      "flex": 1,
                      "fontSize": 14,
                      "paddingLeft": 10,
                    },
                  ]
                }
                testID="text-input"
                value=""
              />
            </View>
          </View>
          <View
            style={
              [
                {
                  "flex": 1,
                  "justifyContent": "center",
                  "marginVertical": 8,
                },
              ]
            }
          >
            <View>
              <Text
                style={
                  [
                    {
                      "color": "#333333",
                      "fontSize": 22,
                      "fontWeight": 500,
                    },
                    [
                      undefined,
                    ],
                  ]
                }
              >
                libraryScreen.featured
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "alignItems": "flex-start",
                    "backgroundColor": "#fff",
                    "borderRadius": 10,
                    "flexDirection": "row",
                    "gap": 16,
                    "marginBottom": 16,
                    "overflow": "hidden",
                  },
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "flex": 1,
                    },
                  ]
                }
              >
                <View>
                  <View
                    collapsable={false}
                    handlerTag={1}
                    handlerType="PanGestureHandler"
                    onGestureHandlerEvent={[Function]}
                    onGestureHandlerStateChange={[Function]}
                    onTouchEnd={[Function]}
                    onTouchStart={[Function]}
                    style={
                      [
                        {
                          "overflow": "hidden",
                        },
                        {
                          "height": 750,
                          "width": 718,
                        },
                        [
                          {
                            "height": 220,
                            "justifyContent": "center",
                            "width": "100%",
                          },
                        ],
                        {
                          "flexDirection": "row",
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        [
                          {
                            "height": 750,
                            "position": "absolute",
                            "width": 718,
                          },
                          {
                            "opacity": undefined,
                            "transform": [
                              {
                                "scale": undefined,
                              },
                              {
                                "translateX": undefined,
                              },
                            ],
                            "zIndex": undefined,
                          },
                        ]
                      }
                      testID="__CAROUSEL_ITEM_0_NOT_READY__"
                    />
                    <View
                      style={
                        [
                          {
                            "height": 750,
                            "position": "absolute",
                            "width": 718,
                          },
                          {
                            "opacity": undefined,
                            "transform": [
                              {
                                "scale": undefined,
                              },
                              {
                                "translateX": undefined,
                              },
                            ],
                            "zIndex": undefined,
                          },
                        ]
                      }
                      testID="__CAROUSEL_ITEM_1_NOT_READY__"
                    />
                    <View
                      style={
                        [
                          {
                            "height": 750,
                            "position": "absolute",
                            "width": 718,
                          },
                          {
                            "opacity": undefined,
                            "transform": [
                              {
                                "scale": undefined,
                              },
                              {
                                "translateX": undefined,
                              },
                            ],
                            "zIndex": undefined,
                          },
                        ]
                      }
                      testID="__CAROUSEL_ITEM_2_NOT_READY__"
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              [
                {
                  "flex": 1,
                  "justifyContent": "center",
                  "marginVertical": 8,
                },
              ]
            }
          >
            <View>
              <Text
                style={
                  [
                    {
                      "color": "#333333",
                      "fontSize": 22,
                      "fontWeight": 500,
                    },
                    [
                      undefined,
                    ],
                  ]
                }
              >
                libraryScreen.popular
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "alignItems": "flex-start",
                    "backgroundColor": "#fff",
                    "borderRadius": 10,
                    "flexDirection": "row",
                    "gap": 16,
                    "marginBottom": 16,
                    "overflow": "hidden",
                  },
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "flex": 1,
                    },
                  ]
                }
              >
                <View>
                  <View
                    collapsable={false}
                    handlerTag={2}
                    handlerType="PanGestureHandler"
                    onGestureHandlerEvent={[Function]}
                    onGestureHandlerStateChange={[Function]}
                    onTouchEnd={[Function]}
                    onTouchStart={[Function]}
                    style={
                      [
                        {
                          "overflow": "hidden",
                        },
                        {
                          "height": 750,
                          "width": 718,
                        },
                        [
                          {
                            "height": 220,
                            "justifyContent": "center",
                            "width": "100%",
                          },
                        ],
                        {
                          "flexDirection": "row",
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        [
                          {
                            "height": 750,
                            "position": "absolute",
                            "width": 718,
                          },
                          {
                            "opacity": undefined,
                            "transform": [
                              {
                                "scale": undefined,
                              },
                              {
                                "translateX": undefined,
                              },
                            ],
                            "zIndex": undefined,
                          },
                        ]
                      }
                      testID="__CAROUSEL_ITEM_0_NOT_READY__"
                    />
                    <View
                      style={
                        [
                          {
                            "height": 750,
                            "position": "absolute",
                            "width": 718,
                          },
                          {
                            "opacity": undefined,
                            "transform": [
                              {
                                "scale": undefined,
                              },
                              {
                                "translateX": undefined,
                              },
                            ],
                            "zIndex": undefined,
                          },
                        ]
                      }
                      testID="__CAROUSEL_ITEM_1_NOT_READY__"
                    />
                    <View
                      style={
                        [
                          {
                            "height": 750,
                            "position": "absolute",
                            "width": 718,
                          },
                          {
                            "opacity": undefined,
                            "transform": [
                              {
                                "scale": undefined,
                              },
                              {
                                "translateX": undefined,
                              },
                            ],
                            "zIndex": undefined,
                          },
                        ]
                      }
                      testID="__CAROUSEL_ITEM_2_NOT_READY__"
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              [
                {
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                },
              ]
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "marginBottom": 16,
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#333333",
                      "fontSize": 22,
                      "fontWeight": 500,
                    },
                    [
                      undefined,
                    ],
                  ]
                }
              >
                libraryScreen.all
              </Text>
              <View>
                <RNGestureHandlerButton
                  collapsable={false}
                  delayLongPress={600}
                  enabled={true}
                  exclusive={true}
                  handlerTag={3}
                  handlerType="NativeViewGestureHandler"
                  onGestureEvent={[Function]}
                  onGestureHandlerEvent={[Function]}
                  onGestureHandlerStateChange={[Function]}
                  onHandlerStateChange={[Function]}
                  rippleColor={0}
                  testID="library"
                  touchSoundDisabled={false}
                >
                  <View
                    accessible={true}
                    collapsable={false}
                    style={
                      {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        [
                          {
                            "backgroundColor": "rgba(255, 153, 0, 0.05)",
                            "borderColor": "#FF9900",
                            "borderRadius": 25,
                            "borderWidth": 2,
                            "paddingHorizontal": 16,
                            "paddingVertical": 8,
                          },
                        ]
                      }
                    >
                      <Text
                        style={
                          [
                            {
                              "color": "#FF9900",
                              "fontSize": 13,
                              "fontWeight": "bold",
                            },
                          ]
                        }
                      >
                        libraryScreen.all
                      </Text>
                    </View>
                  </View>
                </RNGestureHandlerButton>
              </View>
            </View>
            <View
              style={
                [
                  {
                    "flex": 1,
                  },
                ]
              }
            >
              <RCTScrollView
                ItemSeparatorComponent={[Function]}
                contentContainerStyle={
                  {
                    "alignContent": "center",
                    "alignItems": "center",
                    "justifyContent": "space-between",
                    "marginTop": 16,
                  }
                }
                data={[]}
                getItem={[Function]}
                getItemCount={[Function]}
                keyExtractor={[Function]}
                nestedScrollEnabled={false}
                onContentSizeChange={[Function]}
                onLayout={[Function]}
                onMomentumScrollBegin={[Function]}
                onMomentumScrollEnd={[Function]}
                onScroll={[Function]}
                onScrollBeginDrag={[Function]}
                onScrollEndDrag={[Function]}
                removeClippedSubviews={false}
                renderItem={[Function]}
                scrollEventThrottle={50}
                showsVerticalScrollIndicator={false}
                stickyHeaderIndices={[]}
                viewabilityConfigCallbackPairs={[]}
              >
                <View />
              </RCTScrollView>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
